File: src/java/com/kitware/pulse/cdm/patient/actions/SEForcedPause.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package com.kitware.pulse.cdm.patient.actions;

import com.kitware.pulse.cdm.bind.PatientActions.ForcedPauseData;
import com.kitware.pulse.cdm.properties.SEScalarTime;

public class SEForcedPause extends SEConsciousRespirationCommand
{
  protected SEScalarTime period;
  
  public SEForcedPause()
  {
    period = null;
  }
  
  public void copy(SEForcedPause other)
  {
    if(this==other)
      return;

    period = other.period;
  }
  
  @Override
  public SEConsciousRespirationCommand getCopy()
  {
    SEConsciousRespirationCommand command = new SEForcedPause();
    command.copy(this);
    
    return command;
  }
  
  @Override
  public void reset()
  {
    super.reset();

    if (period != null)
      period.invalidate();
  }
  
  @Override
  public boolean isValid()
  {
    return hasPeriod();
  }
  
  public static void load(ForcedPauseData src, SEForcedPause dst)
  {
    //SEConsciousRespirationCommand.load(src.getConsciousRespirationCommand(), dst);
    if(src.hasPeriod())
      SEScalarTime.load(src.getPeriod(),dst.getPeriod());    
  }
  
  public static ForcedPauseData unload(SEForcedPause src)
  {
    ForcedPauseData.Builder dst = ForcedPauseData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  
  protected static void unload(SEForcedPause src, ForcedPauseData.Builder dst)
  {
    //SEConsciousRespirationCommand.unload(src,dst.getConsciousRespirationCommand());
    if (src.hasPeriod())
      dst.setPeriod(SEScalarTime.unload(src.period));
  }
  
  public boolean hasPeriod()
  {
    return period != null;
  }
  public SEScalarTime getPeriod()
  {
    if (period == null)
      period = new SEScalarTime();
    return period;
  }
  
  @Override
  public String toString()
  {
    if (period != null)
    {
      return "Breath Hold"
          + "\n\tPeriod: " + getPeriod();
    }
    else
      return "Action not specified properly";
  }
}

File: src/cpp/cdm/patient/actions/SEForcedPause.cpp
/* Distributed under the Apache License, Version 2.0.
See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SEForcedPause.h"
#include "properties/SEScalarTime.h"
#include "io/protobuf/PBPatientActions.h"

SEForcedPause::SEForcedPause(Logger* logger) : SEConsciousRespirationCommand(logger)
{
  m_Period = nullptr;
}

SEForcedPause::~SEForcedPause()
{
  SAFE_DELETE(m_Period);
}

void SEForcedPause::Clear()
{
  SEConsciousRespirationCommand::Clear();
  INVALIDATE_PROPERTY(m_Period);
}

void SEForcedPause::Copy(const SEForcedPause& src)
{
  PBPatientAction::Copy(src, *this);
}

bool SEForcedPause::IsValid() const
{
  return SEConsciousRespirationCommand::IsValid() && HasPeriod();
}

bool SEForcedPause::IsActive() const
{
  return SEConsciousRespirationCommand::IsActive();
}

bool SEForcedPause::HasPeriod() const
{
  return m_Period == nullptr ? false : m_Period->IsValid();
}
SEScalarTime& SEForcedPause::GetPeriod()
{
  if (m_Period == nullptr)
    m_Period = new SEScalarTime();
  return *m_Period;
}
double SEForcedPause::GetPeriod(const TimeUnit& unit) const
{
  if (m_Period == nullptr)
    return SEScalar::dNaN();
  return m_Period->GetValue(unit);
}

void SEForcedPause::ToString(std::ostream &str) const
{
  str << "Breath Hold";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  str << "\n\tPeriod: "; HasPeriod() ? str << *m_Period : str << "NaN";
  str << std::flush;
}